cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -3207,22 +3207,45 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
 
             // Lower to VM calls when there's no access to SSE4.1.
             let ty = ty.unwrap();
-            let libcall = match (ty, op) {
-                (types::F32, Opcode::Ceil) => LibCall::CeilF32,
-                (types::F64, Opcode::Ceil) => LibCall::CeilF64,
-                (types::F32, Opcode::Floor) => LibCall::FloorF32,
-                (types::F64, Opcode::Floor) => LibCall::FloorF64,
-                (types::F32, Opcode::Nearest) => LibCall::NearestF32,
-                (types::F64, Opcode::Nearest) => LibCall::NearestF64,
-                (types::F32, Opcode::Trunc) => LibCall::TruncF32,
-                (types::F64, Opcode::Trunc) => LibCall::TruncF64,
-                _ => panic!(
-                    "unexpected type/opcode {:?}/{:?} in Ceil/Floor/Nearest/Trunc",
-                    ty, op
-                ),
-            };
-
-            emit_vm_call(ctx, flags, triple, libcall, insn, inputs, outputs)?;
+            if !ty.is_vector() {
+                let libcall = match (op, ty) {
+                    (Opcode::Ceil, types::F32) => LibCall::CeilF32,
+                    (Opcode::Ceil, types::F64) => LibCall::CeilF64,
+                    (Opcode::Floor, types::F32) => LibCall::FloorF32,
+                    (Opcode::Floor, types::F64) => LibCall::FloorF64,
+                    (Opcode::Nearest, types::F32) => LibCall::NearestF32,
+                    (Opcode::Nearest, types::F64) => LibCall::NearestF64,
+                    (Opcode::Trunc, types::F32) => LibCall::TruncF32,
+                    (Opcode::Trunc, types::F64) => LibCall::TruncF64,
```

Comment:
I know this isn't part of the PR, but it is curious that these are implemented as lib calls instead of using `ROUNDSD` and `ROUNDSS`--@bnjbvr, do you know why this is? I wonder if there is some NaN handling there that we should also be doing for the packed version...

Reply:
For reference, here is the NaN paragraph from the manual for `ROUNDPD`:

> The Precision Floating-Point Exception is signaled according to the immediate operand. If any source operand is an SNaN then it will be converted to a QNaN. If DAZ is set to ‘1 then denormals will be converted to zero before rounding.

Reply:
Using ROUNDSD and ROUNDSS is mentioned as a TODO as if it is just missing. I thought these were implemented this way because SSE2 was the minimum supported and because it was faster to implement. I was trying to discern/confirm what the spec wanted beforehand (even in the non-scalar before Ben implemented) and I didn't read in any compatibility with the spec and the line about conversion to QNaN but perhaps Ben has a better idea.

Reply:
Yes, roundss/roundsd are sse4.1 or si, if I remember correctly? Definitely worth adding for the scalar case when the right SSE level is available!

I recall that the wasm spec follows what the hardware does for NaNs for these opcodes, at least in the scalar case.

Reply:
👍 Yes, should be a straight forward follow-up patch. I can submit that.